python/tests/simplify.py

```diff
@@ -291,7 +302,13 @@ def process_parent_edges(self, edges):
             x = self.A_head[edge.child]
             while x is not None:
                 if x.right > edge.left and edge.right > x.left:
-                    y = Segment(max(x.left, edge.left), min(x.right, edge.right), x.node)
+                    if self.filter_unary:
+                        y = Segment(
+                            max(x.left, edge.left), min(x.right, edge.right), x.node)
+                    else:
+                        y = Segment(
+                            max(x.left, edge.left), min(x.right, edge.right),
+                            self.node_id_map[edge.child])
```

Comment:
I don't see why this is working - could you explain? How do we know that `node_id_map` has been defined already for `edge.child`? Why do we need a different case for `filter_unary` and not?

Reply:
> Why do we need a different case for filter_unary and not?

After thinking about it for a bit, I realised that we don't, so I've removed this complication in the latest commit. In a previous version I was using the priority queue of segments in a slightly different way, and the distinction was necessary then. The two cases should produce the same output now.

thanks for thinking about this so carefully @petrelharp!